achievements.lua

```diff
@@ -168,41 +321,202 @@ function achievements.initialize(gamedata, prevent_debug)
+	end
+
+	gfx.pushContext()
+	gfx.setImageDrawMode(gfx.kDrawModeCopy)
+	draw_card_cache[ach.id]:draw(x, y);
+	gfx.popContext()
+end
+
+achievements.drawCard = function (achievement_or_id, x, y)
+	local ach = resolve_achievement_or_id(achievement_or_id)
+	if not ach then
+		achievements.onUnconfigured("attempt to draw unconfigured achievement '" .. achievement_or_id .. "'", 2)
+		return
+	end
+	if x == nil or y == nil then
+		x = achievements.displayGrantedDefaultX
+		y = achievements.displayGrantedDefaultY
+	end
+	draw_card_unsafe(ach, x, y)
+end
+
+local function animate_granted_unsafe(ach, x, y, msec_since_granted, draw_card_func)
+	-- TODO: like for drawing, this needs a bit more care and attention from someone with an art eye
+	draw_card_func(
+		ach,
+		x + 7.0 * math.sin(msec_since_granted / 90.0),
+		y + (msec_since_granted / 10.0)
+	)
+	return msec_since_granted <= achievements.displayGrantedMilliseconds
+end
+
+achievements.animateGranted = function(achievement_or_id, x, y, msec_since_granted, draw_card_func)
+	local ach = resolve_achievement_or_id(achievement_or_id)
+	if not ach then
+		achievements.onUnconfigured("attempt to animate unconfigured achievement '" .. achievement_or_id .. "'", 2)
+		return
+	end
+	if x == nil or y == nil then
+		x = achievements.displayGrantedDefaultX
+		y = achievements.displayGrantedDefaultY
+	end
+	if msec_since_granted == nil then
+		-- for now, the animation will take an equal time in as out, so 'half-time' is a good position to draw unspecified
+		msec_since_granted = achievements.displayGrantedMilliseconds / 2
+	end
+	if draw_card_func == nil then
+		draw_card_func = draw_card_unsafe
+	end
+	return animate_granted_unsafe(ach, x, y, msec_since_granted, draw_card_func)
+end
+
+local animate_coros = {}
+achievements.updateVisuals = function ()
+	for achievement_id, coro_func in pairs(animate_coros) do
+		if not coroutine.resume(coro_func) then
+			animate_coros[achievement_id] = nil
+		end
+	end
+end
+
+local last_grant_display_msec = -achievements.displayGrantedDelayNext
+local function start_granted_animation(ach, draw_card_func, animate_func)
+	draw_card_cache[ach.id] = nil
+	-- tie display-coroutine to achievement-id, so that the system doesn't get confused by rapid grant/revoke
+	animate_coros[ach.id] = coroutine.create(
+		function ()
+			-- NOTE: use getCurrentTimeMilliseconds here (regardless of time granted), since that'll take into account game-pausing.
+			local start_msec = 0
+			repeat
+				start_msec = playdate.getCurrentTimeMilliseconds()
+				coroutine.yield()
+			until start_msec > (last_grant_display_msec + achievements.displayGrantedDelayNext)
+			last_grant_display_msec = start_msec
+			local current_msec = start_msec
+			while animate_func(
+				ach,
+				achievements.displayGrantedDefaultX,
+				achievements.displayGrantedDefaultY,
+				current_msec - start_msec,
+				draw_card_func
+			) do
+				coroutine.yield()
+				current_msec = playdate.getCurrentTimeMilliseconds()
+			end
```

Comment:
This seems a little over-complicated. I don't see any reason that `animate_func` and `draw_card_func` can't be the same function. I do appreciate the use of while to tick the animation function each frame though, that's a nice trick.

Reply:
Done!